### PR TITLE
[DO NOT REVIEW YET]PM Support for BYT/CHT + module load/unload fix for BSW

### PR DESCRIPTION
**Pull request by ranj063**
Needs https://github.com/thesofproject/sof/pull/2137

### CODE REVIEW

**Comment by ranj063** [comment]
@xiulipan @fredoh9 why are all devices failing in this PR and the other PR's?

File: sound/soc/sof/sof-acpi-dev.c [hunk 1]
```diff
@@ -203,6 +208,8 @@ static int sof_acpi_probe(struct platform_device *pdev)
 
 static int sof_acpi_remove(struct platform_device *pdev)
 {
+	pm_runtime_get_noresume(&pdev->dev);
+
```
Comment (plbossart) [inline sound/soc/sof/sof-acpi-dev.c]: why is this needed, and shouldn't it be below the test line 266?

File: sound/soc/sof/sof-acpi-dev.c [hunk 2]
```diff
@@ -133,7 +133,12 @@ static void sof_acpi_probe_complete(struct device *dev)
 	/* allow runtime_pm */
 	pm_runtime_set_autosuspend_delay(dev, SND_SOF_SUSPEND_DELAY_MS);
 	pm_runtime_use_autosuspend(dev);
+	pm_runtime_set_active(dev);
 	pm_runtime_enable(dev);
+	pm_runtime_mark_last_busy(dev);
+
+	/* put_autosuspend doesn't seem to work, so use put_noidle */
+	pm_runtime_put_noidle(dev);
```
Comment (plbossart) [inline sound/soc/sof/sof-acpi-dev.c]: for Soundwire we use a different sequence

		pm_runtime_set_autosuspend_delay(&md->dev, 3000);
		pm_runtime_use_autosuspend(&md->dev);
		pm_runtime_mark_last_busy(&md->dev);

		pm_runtime_set_active(&md->dev);
		pm_runtime_enable(&md->dev);
	
		pm_runtime_idle(&md->dev);

I am not sure what you would do a put_noidle and use a get_noresume, this doesn't look very useful to me to do a put if you haven't done a get first. This is different from PCI where the core does play with refcounts.

File: sound/soc/sof/intel/byt.c [hunk 3]
```diff
@@ -423,6 +423,29 @@ static int byt_remove(struct snd_sof_dev *sdev)
...
+	snd_sof_dsp_update_bits64(sdev, BYT_DSP_BAR, SHIM_IMRX, 0x3, 0x3);
+	snd_sof_dsp_update_bits64(sdev, BYT_DSP_BAR, SHIM_IMRD, 0x3, 0x3);
+
+	/* put DSP into reset, set reset vector */
+	snd_sof_dsp_update_bits64(sdev, BYT_DSP_BAR, SHIM_CSR,
+				  SHIM_BYT_CSR_RST | SHIM_BYT_CSR_VECTOR_SEL,
+				  SHIM_BYT_CSR_RST | SHIM_BYT_CSR_VECTOR_SEL);
+
+	return 0;
+}
```
Comment (plbossart) [inline sound/soc/sof/intel/byt.c:479]: this looks copy-pasted from 

````c
static int byt_remove(struct snd_sof_dev *sdev)
{
	/* disable Interrupt from both sides */
	snd_sof_dsp_update_bits64(sdev, BYT_DSP_BAR, SHIM_IMRX, 0x3, 0x3);
	snd_sof_dsp_update_bits64(sdev, BYT_DSP_BAR, SHIM_IMRD, 0x3, 0x3);

	/* put DSP into reset, set reset vector */
	snd_sof_dsp_update_bits64(sdev, BYT_DSP_BAR, SHIM_CSR,
				  SHIM_BYT_CSR_RST | SHIM_BYT_CSR_VECTOR_SEL,
				  SHIM_BYT_CSR_RST | SHIM_BYT_CSR_VECTOR_SEL);

	return 0;
}
````

maybe use a helper for all this?

File: sound/soc/sof/ipc.c [hunk 4]
```diff
@@ -313,7 +313,8 @@ int snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
 	struct snd_sof_ipc_msg *msg = &sdev->ipc->msg;
 
 	if (msg->ipc_complete) {
-		dev_err(sdev->dev, "error: no reply expected, received 0x%x",
+		dev_dbg(sdev->dev,
+			"no reply expected, received 0x%x, will be ignored",
```
Comment (plbossart) [inline sound/soc/sof/ipc.c:317]: you return -EINVAL so how is it not an error?

**Comment by fredoh9** [comment]
@ranj063 There was a problem in test script. Will fix it now.

File: sound/soc/sof/pm.c [hunk 5]
```diff
@@ -53,8 +53,12 @@ static int sof_resume(struct device *dev, bool runtime_resume)
 	int ret;
 
 	/* do nothing if dsp resume callbacks are not set */
-	if (!sof_ops(sdev)->resume || !sof_ops(sdev)->runtime_resume)
-		return 0;
+	if (runtime_resume)
```
Comment (plbossart) [inline sound/soc/sof/pm.c:56]: This one is a bit controversial

I can see that you could support suspend/resume but not their pm_runtime versions.
would the opposite make any sense, support pm_runtime but not system suspend?


I would reword the commit message. This is not a fix, it's allowing for more granularity in what can be supported. this allows pm_runtime to be unsupported while normal resume is supported.
But I don't know if the other way around makes sense.

File: sound/soc/sof/pm.c [hunk 6]
```diff
@@ -136,6 +136,9 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 		if (!sof_ops(sdev)->suspend)
 			return 0;
 
+	/* remove pipelines */
```
Comment (plbossart) [inline sound/soc/sof/pm.c]: so this one would be a fix of your previous work, and should be bundled with it rather than the byt pm fixes, right?

File: sound/soc/sof/sof-audio.c [hunk 7]
```diff
@@ -53,6 +53,52 @@ int sof_set_hw_params_upon_resume(struct device *dev)
 	return snd_sof_dsp_hw_params_upon_resume(sdev);
 }
 
+int sof_destroy_pipelines(struct device *dev)
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: i am still confused why this needs to be done by hand and isn't done by topology and friends?

File: sound/soc/sof/sof-acpi-dev.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/sof-acpi-dev.c]: @plbossart this one is very confusing for me. I must have played with all the options but for pm_runtime_idle(). let me try that

File: sound/soc/sof/ipc.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/ipc.c:317]: Good point. I should just change this to a void function. we dont really care about the return value anyway

File: sound/soc/sof/pm.c (hunk 5 above)
Comment (ranj063) [reply sound/soc/sof/pm.c:56]: @plbossart this change makes no assumption about whether runtime PM or system PM is supported. sof_resume() and sof_suspend() are called for both the runtime PM and system PM cases. So all I am doing is if we're runtime suspending, check for the platform-specific runtime-suspend callback and so on for the other cases. 
I called it a fix because previously, we didnt check for the appropriate op depending on whether it was runtime callback or system PM callback

File: sound/soc/sof/pm.c (hunk 6 above)
Comment (ranj063) [reply sound/soc/sof/pm.c]: @plbossart yes, it should have been there too but it is all the more necessary for BYT/CHT. What's happening with BYT/CHT is that when performing runtime suspend, the DSP doesnt really get powered off and if I dont destroy the pipelines, runtime resume fails because we restore the pipelines in the resume callback.

File: sound/soc/sof/sof-audio.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: Good idea. Let me move it to topology.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: @plbossart sorry I dont think it can be done by topology. When removing the SOF device, topology can handle this but when dealing with runtime PM, it has be done here.

File: sound/soc/sof/sof-acpi-dev.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/sof-acpi-dev.c]: @plbossart I fixed this. I problems I was seeing what not because of this after all.

File: sound/soc/sof/pm.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/sof/pm.c]: do you mean to say that on BYT/CHT, the firmware still has all the context and when sending the restore pipeline IPCs you get error messages?

It seems like a case of not sending the IPC if they are not needed, instead of destroying the pipelines and having to recreate them again?

Comment (ranj063) [reply sound/soc/sof/pm.c]: @plbossart that is true. But is it acceptable that on BYT/CHT the DSP is not really going to be in D3 when we runtime suspend and will only be done during system suspend?

Comment (plbossart) [reply sound/soc/sof/pm.c]: @ranj063 whether you destroy the pipelines or not, the DSP will not be power gated on BYT/CHT. That's hardware, nothing we can do. In theory clock gating is still there but the leakage can't be controlled by power switches.

 I know that on some Android platforms S0ix was enabled, but there are a ton of dependencies on PMC/BIOS, at this point it's not worth it. The legacy firmware doesn't do anything either for D3.

Comment (plbossart) [reply sound/soc/sof/pm.c]: overall the fix should really be to deal with the topology only at boot and system resume for BYT/CHT, and for other devices in addition for each pm_runtime resume

Comment (ranj063) [reply sound/soc/sof/pm.c]: > I know that on some Android platforms S0ix was enabled, but there are a ton of dependencies on PMC/BIOS, at this point it's not worth it. The legacy firmware doesn't do anything either for D3.

@plbossart OK then should we just not bother enabling runtime PM for BYT/CHT at all?

Comment (ranj063) [reply sound/soc/sof/pm.c]: > overall the fix should really be to deal with the topology only at boot and system resume for BYT/CHT, and for other devices in addition for each pm_runtime resume

@plbossart what happens when we separate the audio client device from the core? We are going to need to take down the pipelines during suspend and restore them during resume right?

Comment (plbossart) [reply sound/soc/sof/pm.c]: > > I know that on some Android platforms S0ix was enabled, but there are a ton of dependencies on PMC/BIOS, at this point it's not worth it. The legacy firmware doesn't do anything either for D3.
> 
> @plbossart OK then should we just not bother enabling runtime PM for BYT/CHT at all?

there's really not much value indeed. No one will see the difference.

Comment (plbossart) [reply sound/soc/sof/pm.c]: > > overall the fix should really be to deal with the topology only at boot and system resume for BYT/CHT, and for other devices in addition for each pm_runtime resume
> 
> @plbossart what happens when we separate the audio client device from the core? We are going to need to take down the pipelines during suspend and restore them during resume right?

it's a good question - as in I don't know the answer.

we have a similar case with SoundWire where sometimes the parent keeps the context and sometimes it doesn't, and we pass this information to the children. So maybe on resume the child device need to be told that all context was lost and that it needs to recreate the whole shebang.

File: sound/soc/sof/pm.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/pm.c:56]: @ranj063 when maintainers see a commit message starting with 'fix' they think it's a real problem that needs to be handled on previous -stable versions. What you are doing is 'split checks for platform-specific callbacks', not a real fix.

File: sound/soc/sof/pm.c [hunk 8]
```diff
@@ -136,6 +136,9 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
...
+	sof_destroy_pipelines(dev);
+
```
Comment (plbossart) [inline sound/soc/sof/pm.c]: i still think this is controversial.

For system suspend-resume, this is not needed. The DSP will lose context and be powered off.

For pm_runtime suspend-resume, this is needed only on Baytrail.

It feels this should really be done conditionally to avoid doing unnecessary things on all platforms where this is useless.

File: sound/soc/sof/pm.c [hunk 9]
```diff
@@ -124,9 +128,13 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
...
-	/* do nothing if dsp suspend callback is not set */
-	if (!sof_ops(sdev)->suspend)
-		return 0;
+	/* do nothing if dsp suspend callbacks are not set */
+	if (runtime_suspend)
+		if (!sof_ops(sdev)->runtime_suspend)
+			return 0;
+	else
+		if (!sof_ops(sdev)->suspend)
+			return 0;
```
Comment (RanderWang) [inline sound/soc/sof/pm.c:137]: @ranj063  Do we need these checks here ? they are already checked in snd_sof_dsp_suspend and snd_sof_dsp_runtime_suspend

**Comment by youling257** [comment]
This series patches cause dsp panic.
[  230.669504] sof-audio-acpi 80860F28:00: error : DSP panic!
[  230.669529] sof-audio-acpi 80860F28:00: panic: dsp_oops_offset 1329152 offset 1329152
[  230.669591] sof-audio-acpi 80860F28:00: ipc tx succeeded: 0x40010000
[  230.669600] sof-audio-acpi 80860F28:00: error: invalid wait state
[  230.669615] sof-audio-acpi 80860F28:00: error: trace point 00000000
[  230.669630] sof-audio-acpi 80860F28:00: error: panic at .../drivers/intel/baytrail/ipc.c0:48
[  230.669643] sof-audio-acpi 80860F28:00: error: DSP Firmware Oops
[  230.669661] sof-audio-acpi 80860F28:00: EXCCAUSE 0x0000003f EXCVADDR 0x00000000 PS       0x00060525 SAR     0x00000000
[  230.669678] sof-audio-acpi 80860F28:00: EPC1     0x00000000 EPC2     0x00000000 EPC3     0x00000000 EPC4    0xff2cf783
[  230.669695] sof-audio-acpi 80860F28:00: EPC5     0xff2cf783 EPC6     0x00000000 EPC7     0x00000000 DEPC    0x00000000
[  230.669741] sof-audio-acpi 80860F28:00: EPS2     0x00000000 EPS3     0x00000000 EPS4     0x00060120 EPS5    0x00060120
[  230.669761] sof-audio-acpi 80860F28:00: EPS6     0x00000000 EPS7     0x00000000 INTENABL 0x00000000 INTERRU 0x000000a2
[  230.669776] sof-audio-acpi 80860F28:00: stack dump from 0xff31c550
[  230.669797] sof-audio-acpi 80860F28:00: 0xff31c550: ff31c570 00000000 00060525 00060320
[  230.669817] sof-audio-acpi 80860F28:00: 0xff31c554: 692f6c69 632e6370 00000030 00000000
[  230.669837] sof-audio-acpi 80860F28:00: 0xff31c558: fea71900 749fe392 00144800 00000000
[  230.669857] sof-audio-acpi 80860F28:00: 0xff31c55c: 80000800 00000000 7b0d0818 ffff9d89
[  230.669876] sof-audio-acpi 80860F28:00: 0xff31c560: 7b0e7e00 ffff9d89 7b12ea00 ffff9d89
[  230.669896] sof-audio-acpi 80860F28:00: 0xff31c564: 9b4c60cc ffffffff 00000004 00000000
[  230.669915] sof-audio-acpi 80860F28:00: 0xff31c568: 7c29af40 ffff9d89 4005bb28 ffffa1f1
[  230.669935] sof-audio-acpi 80860F28:00: 0xff31c56c: 7c29af40 ffff9d89 00000000 00000000
[  230.669953] sof-audio-acpi 80860F28:00: error: ipc host -> DSP: pending no complete no raw 0x0
[  230.669969] sof-audio-acpi 80860F28:00: error: mask host: pending yes complete yes raw 0xffffffffffffffff
[  230.669984] sof-audio-acpi 80860F28:00: error: ipc DSP -> host: pending no complete no raw 0x800008000dead00a
[  230.670000] sof-audio-acpi 80860F28:00: error: mask DSP: pending yes complete yes raw 0xffffffffffffffff

**Comment by plbossart** [comment]
@ranj063 can we update this PR to just have suspend-resume for BYT/CHT (no pm_runtime support) and forget about the load/unload issue with BSW?
Thanks!

**Comment by youling257** [comment]
cherry-pick these five commit with linux kernel master branch, build failed.

sound/soc/sof/intel/byt.c:822:13: error: initialization of ‘int (*)(struct snd_sof_dev *, u32)’ {aka ‘int (*)(struct snd_sof_dev *, unsigned int)’} from incompatible pointer type ‘int (*)(struct snd_sof_dev *)’ [-Werror=incompatible-pointer-types]
  822 |  .suspend = byt_suspend,
      |             ^~~~~~~~~~~
sound/soc/sof/intel/byt.c:822:13: note: (near initialization for ‘sof_byt_ops.suspend’)
sound/soc/sof/intel/byt.c:903:13: error: initialization of ‘int (*)(struct snd_sof_dev *, u32)’ {aka ‘int (*)(struct snd_sof_dev *, unsigned int)’} from incompatible pointer type ‘int (*)(struct snd_sof_dev *)’ [-Werror=incompatible-pointer-types]
  903 |  .suspend = byt_suspend,
      |             ^~~~~~~~~~~
sound/soc/sof/intel/byt.c:903:13: note: (near initialization for ‘sof_cht_ops.suspend’)
cc1: some warnings being treated as errors

**Comment by youling257** [comment]
can you fix pm support for BYT/CHT?

**Comment by eballetbo** [comment]
Fixing the above build issue is simply this:

```
-static int byt_suspend(struct snd_sof_dev *sdev)
+static int byt_suspend(struct snd_sof_dev *sdev, u32 target_state)
```
Although the patches apply on current mainline (with the above fix), suspend/resume doesn't seem to work:
```
[ 1299.061192] sof-audio-acpi 808622A8:00: error: ipc timed out for 0x60010000 size 108
[ 1299.061255] sof-audio-acpi 808622A8:00: invalid header size 0x2088002. FW oops is bogus                                                                    
[ 1299.061272] sof-audio-acpi 808622A8:00: error: unexpected fault 0x00000000 trace 0x00000000                                                                
[ 1299.061290] sof-audio-acpi 808622A8:00: error: ipc host -> DSP: pending no complete no raw 0x8000000000000000                                              
[ 1299.061303] sof-audio-acpi 808622A8:00: error: mask host: pending no complete no raw 0x0                                                                   
[ 1299.061315] sof-audio-acpi 808622A8:00: error: ipc DSP -> host: pending no complete no raw 0x0                                                             
[ 1299.061328] sof-audio-acpi 808622A8:00: error: mask DSP: pending no complete no raw 0x0 
[ 1299.061343] sof-audio-acpi 808622A8:00: error: hw params ipc failed for stream 0                                                                           
[ 1299.061358] sof-audio-acpi 808622A8:00: ASoC: 808622A8:00 hw params failed: -110                                                                           
[ 1299.061374]  Low Latency: ASoC: hw_params FE failed -110
```

**Comment by plbossart** [comment]
replaced by PR #2138
